### PR TITLE
Drop Ruby 1.9, switch to using Module.prepend instead of method aliasing

### LINKED ISSUE #603: Use Module.prepend for new instrumentation
Fixed in https://github.com/appsignal/appsignal-ruby/pull/683

## Background
The issue reported in #583, causing SystemStackError errors, is because the `http_logger` gem uses `Module.prepend` to monkeypatch the Net::HTTP standard library. This is incompatible with the "method aliasing" method we use in the AppSignal Ruby gem to monkeypatch methods.

We updated the monkeypatch method for the Net::HTTP integration in #597 to use `Module.prepend` for newer Ruby versions, while supporting Ruby 1.9. This was released in Ruby gem version 2.10.5.

The fix in #597 was reverted in PR #604 and released in Ruby gem 2.10.7.

## The problem

However, using the `Module.prepend` method to monkeypatch, breaks compatibility with other libraries that use the "method aliasing" method, such as New Relic, Skylight and Airbrake, as reported in the comments in https://github.com/appsignal/appsignal-ruby/pull/597#issuecomment-622379707 and https://github.com/appsignal/support/issues/71 (private issue)

We probably can't support every method now and in the future, but I think we should move towards the `Module.prepend` solution rather than "method aliasing" if at all possible.

## Potential solution

To support Ruby 1.9 and maintain compatibility with other gems we could introduce a "legacy/compatibility mode" that uses the "method aliasing" method.

For example:

```yaml
# config/appsignal.yml
default: &defaults
  legacy_mode: true

production:
  <<: *defaults
```

```ruby
# lib/appsignal/hooks/net_http.rb
def install
  if Appsignal.legacy_mode? # For Ruby 1.9 and `legacy_mode: true`
    Net::HTTP.class_eval do
      # Method aliasing
      alias request_without_appsignal request

      def request(...)
        Appsignal.instrument(...) do
          request_without_appsignal(...)
        end
      end
    end
  else
    # Module.prepend
    require "appsignal/integrations/net_http"
    Net::HTTP.send(:prepend, Appsignal::Integrations::NetHttpIntegration)
  end
end
```

### PR DESCRIPTION
**Pull request by jeffkreeftmeijer**
Part of Appsignal for Ruby 3.0. Closes #603.

### CODE REVIEW

File: lib/appsignal/integrations/redis.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+module Appsignal
+  module Integrations
+    module RedisIntegration
+      def process(commands, &block)
+        sanitized_commands = commands.map do |command, *args|
+          "#{command}#{" ?" * args.size}"
+        end.join("\n")
+
+        Appsignal.instrument "query.redis", id, sanitized_commands do
+          process_without_appsignal(commands, &block)
```
Comment (tombruijn) [inline lib/appsignal/integrations/redis.rb:12]:
```suggestion
          super
```

`process_without_appsignal` doesn't exist anymore.

**Review by tombruijn — CHANGES_REQUESTED** [review]
Found a couple of things. I haven't manually tested this yet.

**Review by matsimitsu — APPROVED** [review]

**Comment by xadamy** [comment]
Removing myself from the reviewer list for this one, as it's unlikely I'll get round to reviewing it before I go on leave, unfortunately.

**Comment by tombruijn** [comment]
@jeffkreeftmeijer can I help to get this merged? Fix the tests and such.

**Comment by jeffkreeftmeijer** [comment]
@tombruijn that’d be great indeed!

**Comment by tombruijn** [comment]
I continued this in #683 and only needs the fixes I suggested here in this pr before review.

**Comment by tombruijn** [comment]
Closing this in favor of #683

**Pull request closed without merging**